_posts/2015-04-24-FFI-and-Rust.md

```diff
@@ -0,0 +1,264 @@
+
+Next we've got our function definition, and the most interesting part about
+this is the keyword `extern`. This is a specialized form of specifying the [ABI
+for a function][abi-fn] which enables C to know how to call this function.
+
+[abi-fn]: http://doc.rust-lang.org/reference.html#extern-functions
+
+Finally, if you [take a look at the `Cargo.toml`][cargo-toml] you'll see that
+this library is not compiled as a normal Rust library (rlib) but instead as a
+static archive called a staticlib. This enables all the relevant Rust code to be
+linked statically into the C program we're about to produce.
+
+[cargo-toml]: https://github.com/alexcrichton/rust-ffi-examples/blob/master/c-to-rust/Cargo.toml#L8
+
+Now that we've got our Rust library squared away, let's write our C program
+which will call Rust.
+
+```c
+#include <stdint.h>
+#include <stdio.h>
+
+extern int32_t double_input(int32_t input);
+
+int main() {
+    int input = 4;
+    int output = double_input(input);
+    printf("%d * 2 = %d\n", input, output);
+    return 0;
+}
+```
+
+Here we can see that C, like Rust, needs to declare the `double_input` function
+that Rust defined. Other than that though everything is ready to go! If you run
+`make` from the [directory on GitHub][c2rust] you'll see these examples getting
+compiled and linked together and the final executable should run and print
+`4 * 2 = 8`.
+
+Rust's lack of a garbage collector and runtime enables this seamless transition
+from C to Rust. The external C code does not need to perform any setup on Rust's
+behalf, making the transition that much cheaper.
+
+### Beyond C
+
+Up to now we've seen how FFI in Rust has 0 overhead and how we can use Rust's
+concept of ownership to write safe bindings to C libraries. If you're not using
+C, however, you're still in luck! These features of Rust enable it to also be
+called from [Python][py2rust], [Ruby][rb2rust], [Javascript][js2rust], and many
+more languages.
+
+[py2rust]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/python-to-rust
+[rb2rust]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/ruby-to-rust
+[js2rust]: https://github.com/alexcrichton/rust-ffi-examples/tree/master/node-to-rust
+
+A common desire for writing C code in these languages is to speed up some
+component of a library or application that's performance critical. With the
+features of Rust we've seen here, however, Rust is just as suitable for this
+sort of usage. One of Rust's first production users,
+[Skylight](https://www.skylight.io), is a great example of the benefits that
+Rust can bring even to existing applications by improving the performance and
+memory usage of their data collection agent almost instantly by just using Rust.
+
+Many programmers in languages like Python and Ruby have little systems
+programming experience, so reaching down to C is often avoided as it's tough to
+ensure that the program won't crash in a difficult-to-debug way. Rust, however,
```

Comment:
You might avoid ascribing any particular behavior or thinking to large groups of developers. This can probably be phrased in a way that is less accusatory of the deficiencies of Python and Ruby programs - integrating C is just hard and Rust makes it easier.